File: codeforces/0113B_-_Petr#/main.py
from sys import stdin, stdout
def main():
    BASE = 37
    MOD = 1e7 + 7
    s = stdin.readline().strip()
    n = len(s)
    begin = stdin.readline().strip()
    end = stdin.readline().strip()
    b = set()
    for i in range(0, n):
        k = 0
        while k + i < n and k < len(begin) and s[k + i] == begin[k]:
            k += 1
        if k == len(begin):
            b.add(i)
    e = set()
    for i in range(0, n):
        k = 0
        while k + i < n and k < len(end) and s[k + i] == end[k]:
            k += 1
        if k == len(end):
            e.add(i + k - 1)

    h = [ord(s[0])]
    p = [1]
    for i in range(1, n):
        h.append(int(((h[i - 1] * BASE) % MOD + ord(s[i])) % MOD))
        p.append(int((p[i - 1] * BASE) % MOD))
    se = set()
    for i in b:
        for j in e:
            if j < i + max(len(begin), len(end)) - 1:
                continue
            if i == 0:
                se.add(h[j])
            else:
                se.add((h[j] - int(((h[i - 1] * p[j - i + 1]) % MOD))) % MOD)
    return len(se)

stdout.write('{}\n'.format(main()))

File: codeforces/0113B_-_Petr#/main.cpp
#include <bits/stdc++.h>

using namespace std;

const int MOD = 1e9 + 7;
const int BASE = 37;
const int MAX = 1e7;

typedef long long int ll;

ll h[MAX];
ll p[MAX];

int main() {
  string s, begin, end;
  cin >> s;
  cin >> begin;
  cin >> end;
  unordered_set<int> b;
  for (int i = 0; i < s.size(); i++) {
    int k;
    for (k = 0; s[k + i] && begin[k] && s[k + i] == begin[k]; ++k);
    if (k == begin.size())
      b.insert(i);
  }

  unordered_set<int> e;
  for (int i = 0; i < s.size(); i++) {
    int k;
    for (k = 0; s[k + i] && end[k] && s[k + i] == end[k]; ++k);
    if (k == end.size())
      e.insert(k + i - 1);
  }

  h[0] = s[0];
  p[0] = 1LL;

  for (int i = 1; i < s.size(); i++) {
    h[i] = (h[i - 1] * BASE) + s[i];
    p[i] = p[i - 1] * BASE;
  }

  unordered_set<ll> se;
  for (int i = 0; i < s.size(); i++) {
    if (b.find(i) == b.end()) continue;
    for (int j = i + max(begin.size(), end.size()) - 1; 
             j < s.size(); ++j) {
      if (e.find(j) != e.end()) {
        if (i == 0)
          se.insert(h[j]);
        else
          se.insert(h[j] - (h[i - 1] * p[j - i + 1]));
      }
    }
  }

  cout << se.size() << "\n";

  return 0;
}
